client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt

```diff
@@ -0,0 +1,411 @@
+package app.cash.backfila.client.misk.base
+
+import app.cash.backfila.client.misk.Backfill
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeRequest
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse.Batch
+import app.cash.backfila.protos.clientservice.KeyRange
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse.Instance
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import app.cash.backfila.protos.clientservice.RunBatchResponse
+import com.google.common.base.Preconditions.checkArgument
+import com.google.common.base.Stopwatch
+import com.google.common.collect.ImmutableList
+import com.google.inject.TypeLiteral
+import misk.exceptions.BadRequestException
+import misk.hibernate.DbEntity
+import misk.hibernate.Id
+import misk.hibernate.Operator.GE
+import misk.hibernate.Operator.GT
+import misk.hibernate.Operator.LE
+import misk.hibernate.Query
+import misk.hibernate.Session
+import misk.logging.getLogger
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import java.util.ArrayList
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.persistence.Table
+import kotlin.reflect.KClass
+
+/**
+ * Generic backfill that runs over Hibernate 4 entities.
```

Comment:
are you planning to abstract some of this to share with square internal code?

Reply:
We will see. Right now I'm focused on only getting the kotlin one right. Then I'll see what can be ported over.